packages/dd-trace/src/appsec/iast/analyzers/nosql-injection-mongodb-analyzer.js

```diff
@@ -162,4 +165,25 @@ class NosqlInjectionMongodbAnalyzer extends InjectionAnalyzer {
   }
 }
 
+function iterateMongodbQueryStrings (target, fn, levelKeys = [], depth = 10, visited = new Set()) {
+  if (target !== null && typeof target === 'object') {
+    if (visited.has(target)) return
+
+    visited.add(target)
```

Comment:
While a set seems the obvious thing to use here to limit the number of lookups, it's likely faster to use an array. The reason is that the default lookup will normally not reach a depth of 10. There is probably a break even point somewhere in-between and it might be interesting to check for that.

Allocating a set and creating the hash is just more expensive than allocating an array and iterating over very few entries.

This will however only work in case we also release the values when moving up. Do we actually have to match all former entries or only only ones that are available in the current leaf?

Reply:
You mean using `[].includes(target)` instead of `Set.has(target)`? Yeah, it should work also, I dunno if the difference will be noticed, but I think that we can change it...

> This will however only work in case we also release the values when moving up. Do we actually have to match all former entries or only only ones that are available in the current leaf?

I'm not sure if I understand, the goal of this is to prevent circular dependencies, but also to prevent  checking the same twice, with objects like:

```
{
  a: objectA,
  b: {
      a: objectA
  }
}
```
We don't need to iterate again in `b.a: objectA`